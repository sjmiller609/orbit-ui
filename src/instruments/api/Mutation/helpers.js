const has = (str, find) => ~str.toLowerCase().indexOf(find.toLowerCase())

export const formErrors = ({ error, errors }) => {
  if (!error) return

  let err = JSON.stringify(error).toLowerCase()
  return errors.find(e => {
    if (e.key && has(err, e.key)) return e
    else if (e.allKeys && e.allKeys.every(k => has(err, k))) return e
    else if (e.someKeys && e.someKeys.some(k => has(err, k))) return e
  })
}
