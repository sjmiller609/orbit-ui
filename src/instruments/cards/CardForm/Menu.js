'use strict'
import React from 'react'
import PropTypes from 'prop-types'

import { MenuList, Item } from 'instruments'

import s from './styles.scss'

const Menu = ({ menu }) => {
  return (
    <MenuList className={s.menu}>
      {menu.map(m => (
        <Item key={m.id} to={'#' + m.id}>
          {m.text}
        </Item>
      ))}
    </MenuList>
  )
}

Menu.propTypes = {
  menu: PropTypes.array,
}

export default Menu
