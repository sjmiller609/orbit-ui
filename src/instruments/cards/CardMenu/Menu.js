'use strict'
import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import { Card, MenuList, Item } from 'instruments'

import s from './styles.scss'

const Menu = ({ menu, active, title, id, scrollTo, menuList, className }) => {
  return (
    <Card id={id} title={title} className={classnames(s.menu, className)}>
      <MenuList {...menuList}>
        {menu.map((m, i) => (
          <React.Fragment key={m.id + i.toString()}>
            {m.newForm && <div className={s.border} />}
            <Item
              to={m.to}
              onClick={m.id ? () => scrollTo(m.id) : null}
              active={active === m.id}>
              {m.text}
            </Item>
          </React.Fragment>
        ))}
      </MenuList>
    </Card>
  )
}

Menu.propTypes = {
  active: PropTypes.string,
  id: PropTypes.string,
  title: PropTypes.string,
  className: PropTypes.string,
  menu: PropTypes.array,
  scrollTo: PropTypes.func,
  menuList: PropTypes.object,
}

export default Menu
