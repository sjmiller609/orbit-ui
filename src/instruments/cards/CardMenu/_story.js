import React from 'react'
import { storiesOf } from '@storybook/react'
import { CardMenu, Menu } from 'instruments'

storiesOf('Cards', module).add('Menu', () => <Menu />)
