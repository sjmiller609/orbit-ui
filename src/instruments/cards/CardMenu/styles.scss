@import 'src/styles/mixins.scss';

.menu {
  &.sticky {
    position: sticky;
    top: 16px;
  }
  // matches this.disable value set in index.js
  @media screen and (max-width: 850px) {
    display: none;
  }
}
.cards {
  flex-grow: 1;
}

.border {
  height: 2px;
  background: $lightGray;
  width: 100%;
}
