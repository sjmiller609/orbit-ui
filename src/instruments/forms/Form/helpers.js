// faltten and stringify nested keys
export const unpack = (obj, ns, hash) => {
  if (!obj) return
  return Object.entries(obj).reduce((acc, [key, value]) => {
    const keypath = ns ? `${ns}.${key}` : key
    if (value && typeof value === 'object') {
      return unpack(value, keypath, acc)
    }
    acc[keypath] = value
    return acc
  }, hash || {})
}

const roll = (keys, obj, value) => {
  const key = keys[0]
  // case 1 - not obj
  if (keys.length === 1) {
    obj[key] = value
    return obj
  }
  // case 2 - doesn't exist
  if (!obj[key]) {
    obj[key] = isNaN(keys[1]) ? {} : []
  }
  // case 3 - exists
  obj[key] = roll(
    keys.slice(1),
    Array.isArray(obj[key]) ? [...obj[key]] : { ...obj[key] },
    value
  )

  return obj
}

// convert back into obj
export const pack = (obj, type = {}) => {
  return Object.entries(obj).reduce((acc, [key, value]) => {
    const keys = key.split('.')

    return roll(keys, acc, value)
  }, type)
}

export const packChild = ({ name, obj }) => {
  const obj2 = {}
  let isArray
  Object.entries(obj).forEach(k => {
    if (name.length < k[0].length && k[0].indexOf(name) === 0) {
      const n = k[0].slice(name.length + 1)
      // must be true for all matching
      isArray = !isNaN(n.slice(0, n.indexOf('.')))
      obj2[n] = k[1]
    }
  })
  if (Object.keys(obj2).length) return pack(obj2, isArray ? [] : {})
  return null
}

export const removeChild = ({ name, obj = {} }) => {
  const obj2 = {}
  Object.entries(obj).forEach(k => {
    if (name.length >= k[0].length || k[0].indexOf(name) !== 0) {
      obj2[k[0]] = k[1]
    }
  })
  return obj2
}
