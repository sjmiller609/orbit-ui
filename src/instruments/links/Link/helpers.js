// check if url is external
export const externalUrl = url => {
  if (typeof url === 'object') return false
  const path = /^https?:\/\/|^\/\//i
  if (path.test(url) && !~url.indexOf(window.location.origin)) return true
  return
}
