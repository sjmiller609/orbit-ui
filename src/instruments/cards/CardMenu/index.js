'use strict'
import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import { Row } from 'instruments'
import Menu from './Menu'
import s from './styles.scss'

class CardMenu extends React.Component {
  pos = 0
  scroll = this.scroll.bind(this)
  scrollTo = this.scrollTo.bind(this)
  disable = !this.props.menu.length
  menu = []
  state = {
    focus: this.props.menu.length && this.props.menu[0].id,
    sticky: null,
  }

  componentDidMount() {
    if (window.innerWidth <= 850) this.disable = true
    if (!this.disable) window.addEventListener('scroll', this.scroll)
  }
  componentWillUnmount() {
    if (!this.disable) window.removeEventListener('scroll', this.scroll)
  }

  scroll() {
    if (this.disable) return
    const { menu } = this.props
    const scroll = window.scrollY
    const scroll0 = this.pos

    this.pos = scroll
    const diff = scroll - scroll0
    if (Math.abs(diff) < 5) return

    const menu2 = Array.from(menu)
    if (diff > 0) menu2.reverse()
    const mid = window.innerHeight * 0.55

    const focus = menu2.some(m => this.getFocus(m.id, mid, diff < 0))
    // test for menu sticky scroll position
    const sticky = this.sticky()

    const set = {}
    if (!focus) set.focus = null
    if (sticky !== this.state.sticky) set.sticky = sticky
    if (Object.keys(set).length) this.setState(set)
  }

  getFocus(id, mid, up) {
    const el = document.getElementById(id)
    if (!el) return
    const rect = el.getBoundingClientRect()

    if ((!up && rect.y < mid) || (up && rect.y + rect.height > mid)) {
      this.setState({ focus: id })
      return true
    }
  }
  sticky() {
    const { id } = this.props
    const el = document.getElementById(id)
    if (!el) return
    const rect = el.getBoundingClientRect()
    if (rect.y <= 16) return true
  }
  scrollTo(id) {
    const el = document.getElementById(id)
    el.scrollIntoView()
  }

  render() {
    const { children, id, title, menu, menuList, className } = this.props
    const { focus, sticky } = this.state
    if (this.disable) return children
    return (
      <Row align="flex-start" className={classnames(s.row, className)}>
        <Menu
          id={id}
          menu={menu}
          menuList={menuList}
          title={title}
          active={focus}
          scrollTo={this.scrollTo}
          className={classnames(sticky && s.sticky)}
        />

        <div className={s.cards}>{children}</div>
      </Row>
    )
  }
}

CardMenu.propTypes = {
  title: PropTypes.string,
  id: PropTypes.string,
  children: PropTypes.oneOfType([PropTypes.array, PropTypes.element]),
  className: PropTypes.string,
  menu: PropTypes.array,
  menuList: PropTypes.object,
}

CardMenu.defaultProps = {
  id: 'cardMenu',
  menu: [],
}

export default CardMenu
