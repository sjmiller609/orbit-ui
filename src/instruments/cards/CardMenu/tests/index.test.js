import React from 'react'
import { shallow } from 'enzyme'
import { mountWrap } from 'helpers/tests'
import CardMenu from '../index'
import Menu from '../Menu'

const menu = [
  {
    text: 'API Key',
    id: 'apiKey',
  },
  {
    text: 'Configure',
    id: 'configure',
  },
  {
    text: 'Delete',
    id: 'delete',
    newForm: true,
  },
]

const menuList = {
  button: {
    text: 'Back',
    backArrow: 'arrow',
    style: 'blue',
    to: '/',
  },
}

const props = {
  title: 'Test',
  id: '1234',
  children: ['<p>This is a test</p>'],
  className: 'testClassString',
  menu,
  menuList,
}

const menuProps = {
  active: 'apiKey',
  id: '123',
  title: 'Test',
  className: 'testClassString',
  menu,
  menuList,
}

const propsNoScroll = {
  ...props,
  menu: [],
}

describe('CardMenu', () => {
  let wrapper
  let menu

  beforeEach(() => {
    wrapper = mountWrap(<CardMenu {...props} />)
    menu = shallow(<Menu {...menuProps} />)
  })

  it('should be defined', () => {
    expect(CardMenu).toBeDefined()
    expect(Menu).toBeDefined()
  })

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot()
    expect(menu).toMatchSnapshot()
  })

  it('should handle sticky', () => {
    wrapper.setState({ sticky: true })
    expect(wrapper).toMatchSnapshot()
  })

  it('should handle scroll disabled', () => {
    wrapper = mountWrap(<CardMenu {...propsNoScroll} />)
    expect(wrapper).toMatchSnapshot()
  })
})
