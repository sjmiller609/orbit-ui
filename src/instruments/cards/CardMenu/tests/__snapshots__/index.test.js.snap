// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardMenu should handle scroll disabled 1`] = `
<CardMenu
  className="testClassString"
  id="1234"
  menu={Array []}
  menuList={
    Object {
      "button": Object {
        "backArrow": "arrow",
        "style": "blue",
        "text": "Back",
        "to": "/",
      },
    }
  }
  title="Test"
>
  &lt;p&gt;This is a test&lt;/p&gt;
</CardMenu>
`;

exports[`CardMenu should handle sticky 1`] = `
<CardMenu
  className="testClassString"
  id="1234"
  menu={
    Array [
      Object {
        "id": "apiKey",
        "text": "API Key",
      },
      Object {
        "id": "configure",
        "text": "Configure",
      },
      Object {
        "id": "delete",
        "newForm": true,
        "text": "Delete",
      },
    ]
  }
  menuList={
    Object {
      "button": Object {
        "backArrow": "arrow",
        "style": "blue",
        "text": "Back",
        "to": "/",
      },
    }
  }
  title="Test"
>
  <Row
    align="flex-start"
    className="row testClassString"
  >
    <Flex
      align="flex-start"
      className="row testClassString"
      flow="row"
    >
      <div
        className="flex row undefined align-flex-start row testClassString"
      >
        <Menu
          active="apiKey"
          className="sticky"
          id="1234"
          menu={
            Array [
              Object {
                "id": "apiKey",
                "text": "API Key",
              },
              Object {
                "id": "configure",
                "text": "Configure",
              },
              Object {
                "id": "delete",
                "newForm": true,
                "text": "Delete",
              },
            ]
          }
          menuList={
            Object {
              "button": Object {
                "backArrow": "arrow",
                "style": "blue",
                "text": "Back",
                "to": "/",
              },
            }
          }
          scrollTo={[Function]}
          title="Test"
        >
          <Card
            className="menu sticky"
            id="1234"
            title="Test"
          >
            <Box
              className="card menu sticky"
              id="1234"
              justify="flex-start"
            >
              <Flex
                className="card menu sticky"
                flow="box"
                id="1234"
                justify="flex-start"
              >
                <div
                  className="flex box flex-start card menu sticky"
                  id="1234"
                >
                  <main
                    className="full"
                  >
                    <MenuList
                      button={
                        Object {
                          "backArrow": "arrow",
                          "style": "blue",
                          "text": "Back",
                          "to": "/",
                        }
                      }
                    >
                      <ul
                        className="list"
                      >
                        <Item
                          active={true}
                          exact={true}
                          onClick={[Function]}
                        >
                          <Link
                            activeClassName="active"
                            className="item active"
                            exact={true}
                            onClick={[Function]}
                            title="API Key"
                          >
                            <A
                              activeClassName="active"
                              className="link item active"
                              exact={true}
                              onClick={[Function]}
                              title="API Key"
                            >
                              <a
                                className="link item active"
                                onClick={[Function]}
                                title="API Key"
                              >
                                API Key
                              </a>
                            </A>
                          </Link>
                        </Item>
                        <Item
                          active={false}
                          exact={true}
                          onClick={[Function]}
                        >
                          <Link
                            activeClassName="active"
                            className="item"
                            exact={true}
                            onClick={[Function]}
                            title="Configure"
                          >
                            <A
                              activeClassName="active"
                              className="link item"
                              exact={true}
                              onClick={[Function]}
                              title="Configure"
                            >
                              <a
                                className="link item"
                                onClick={[Function]}
                                title="Configure"
                              >
                                Configure
                              </a>
                            </A>
                          </Link>
                        </Item>
                        <div
                          className="border"
                        />
                        <Item
                          active={false}
                          exact={true}
                          onClick={[Function]}
                        >
                          <Link
                            activeClassName="active"
                            className="item"
                            exact={true}
                            onClick={[Function]}
                            title="Delete"
                          >
                            <A
                              activeClassName="active"
                              className="link item"
                              exact={true}
                              onClick={[Function]}
                              title="Delete"
                            >
                              <a
                                className="link item"
                                onClick={[Function]}
                                title="Delete"
                              >
                                Delete
                              </a>
                            </A>
                          </Link>
                        </Item>
                        <TextButton
                          backArrow="arrow"
                          className="button"
                          style="blue"
                          text="Back"
                          to="/"
                        >
                          <Link
                            backArrow="arrow"
                            className="link blue button"
                            text="Back"
                            to="/"
                          >
                            <Link
                              className="link link blue button"
                              replace={false}
                              text="Back"
                              to="/"
                            >
                              <a
                                className="link link blue button"
                                href="/"
                                onClick={[Function]}
                                text="Back"
                              >
                                <Icon
                                  className="backArrow"
                                  icon="arrow"
                                >
                                  <Imported
                                    className="backArrow"
                                  >
                                    <ReactImportedComponent
                                      className="backArrow"
                                      loadable={
                                        Object {
                                          "done": true,
                                          "importFunction": [Function],
                                          "load": [Function],
                                          "mark": Array [],
                                          "ok": true,
                                          "payload": Object {
                                            "process": [Function],
                                          },
                                          "promise": Promise {},
                                          "reset": [Function],
                                        }
                                      }
                                      render={[Function]}
                                    >
                                      <UnconnectedReactImportedComponent
                                        async={false}
                                        className="backArrow"
                                        exportPicker={[Function]}
                                        loadable={
                                          Object {
                                            "done": true,
                                            "importFunction": [Function],
                                            "load": [Function],
                                            "mark": Array [],
                                            "ok": true,
                                            "payload": Object {
                                              "process": [Function],
                                            },
                                            "promise": Promise {},
                                            "reset": [Function],
                                          }
                                        }
                                        render={[Function]}
                                        streamId={0}
                                      >
                                        <img
                                          className="backArrow"
                                          src={
                                            Object {
                                              "process": [Function],
                                            }
                                          }
                                        />
                                      </UnconnectedReactImportedComponent>
                                    </ReactImportedComponent>
                                  </Imported>
                                </Icon>
                                Back
                              </a>
                            </Link>
                          </Link>
                        </TextButton>
                      </ul>
                    </MenuList>
                  </main>
                </div>
              </Flex>
            </Box>
          </Card>
        </Menu>
        <div
          className="cards"
        >
          &lt;p&gt;This is a test&lt;/p&gt;
        </div>
      </div>
    </Flex>
  </Row>
</CardMenu>
`;

exports[`CardMenu should render correctly 1`] = `
<CardMenu
  className="testClassString"
  id="1234"
  menu={
    Array [
      Object {
        "id": "apiKey",
        "text": "API Key",
      },
      Object {
        "id": "configure",
        "text": "Configure",
      },
      Object {
        "id": "delete",
        "newForm": true,
        "text": "Delete",
      },
    ]
  }
  menuList={
    Object {
      "button": Object {
        "backArrow": "arrow",
        "style": "blue",
        "text": "Back",
        "to": "/",
      },
    }
  }
  title="Test"
>
  <Row
    align="flex-start"
    className="row testClassString"
  >
    <Flex
      align="flex-start"
      className="row testClassString"
      flow="row"
    >
      <div
        className="flex row undefined align-flex-start row testClassString"
      >
        <Menu
          active="apiKey"
          className=""
          id="1234"
          menu={
            Array [
              Object {
                "id": "apiKey",
                "text": "API Key",
              },
              Object {
                "id": "configure",
                "text": "Configure",
              },
              Object {
                "id": "delete",
                "newForm": true,
                "text": "Delete",
              },
            ]
          }
          menuList={
            Object {
              "button": Object {
                "backArrow": "arrow",
                "style": "blue",
                "text": "Back",
                "to": "/",
              },
            }
          }
          scrollTo={[Function]}
          title="Test"
        >
          <Card
            className="menu"
            id="1234"
            title="Test"
          >
            <Box
              className="card menu"
              id="1234"
              justify="flex-start"
            >
              <Flex
                className="card menu"
                flow="box"
                id="1234"
                justify="flex-start"
              >
                <div
                  className="flex box flex-start card menu"
                  id="1234"
                >
                  <main
                    className="full"
                  >
                    <MenuList
                      button={
                        Object {
                          "backArrow": "arrow",
                          "style": "blue",
                          "text": "Back",
                          "to": "/",
                        }
                      }
                    >
                      <ul
                        className="list"
                      >
                        <Item
                          active={true}
                          exact={true}
                          onClick={[Function]}
                        >
                          <Link
                            activeClassName="active"
                            className="item active"
                            exact={true}
                            onClick={[Function]}
                            title="API Key"
                          >
                            <A
                              activeClassName="active"
                              className="link item active"
                              exact={true}
                              onClick={[Function]}
                              title="API Key"
                            >
                              <a
                                className="link item active"
                                onClick={[Function]}
                                title="API Key"
                              >
                                API Key
                              </a>
                            </A>
                          </Link>
                        </Item>
                        <Item
                          active={false}
                          exact={true}
                          onClick={[Function]}
                        >
                          <Link
                            activeClassName="active"
                            className="item"
                            exact={true}
                            onClick={[Function]}
                            title="Configure"
                          >
                            <A
                              activeClassName="active"
                              className="link item"
                              exact={true}
                              onClick={[Function]}
                              title="Configure"
                            >
                              <a
                                className="link item"
                                onClick={[Function]}
                                title="Configure"
                              >
                                Configure
                              </a>
                            </A>
                          </Link>
                        </Item>
                        <div
                          className="border"
                        />
                        <Item
                          active={false}
                          exact={true}
                          onClick={[Function]}
                        >
                          <Link
                            activeClassName="active"
                            className="item"
                            exact={true}
                            onClick={[Function]}
                            title="Delete"
                          >
                            <A
                              activeClassName="active"
                              className="link item"
                              exact={true}
                              onClick={[Function]}
                              title="Delete"
                            >
                              <a
                                className="link item"
                                onClick={[Function]}
                                title="Delete"
                              >
                                Delete
                              </a>
                            </A>
                          </Link>
                        </Item>
                        <TextButton
                          backArrow="arrow"
                          className="button"
                          style="blue"
                          text="Back"
                          to="/"
                        >
                          <Link
                            backArrow="arrow"
                            className="link blue button"
                            text="Back"
                            to="/"
                          >
                            <Link
                              className="link link blue button"
                              replace={false}
                              text="Back"
                              to="/"
                            >
                              <a
                                className="link link blue button"
                                href="/"
                                onClick={[Function]}
                                text="Back"
                              >
                                <Icon
                                  className="backArrow"
                                  icon="arrow"
                                >
                                  <Imported
                                    className="backArrow"
                                  >
                                    <ReactImportedComponent
                                      className="backArrow"
                                      loadable={
                                        Object {
                                          "done": true,
                                          "importFunction": [Function],
                                          "load": [Function],
                                          "mark": Array [],
                                          "ok": true,
                                          "payload": Object {
                                            "process": [Function],
                                          },
                                          "promise": Promise {},
                                          "reset": [Function],
                                        }
                                      }
                                      render={[Function]}
                                    >
                                      <UnconnectedReactImportedComponent
                                        async={false}
                                        className="backArrow"
                                        exportPicker={[Function]}
                                        loadable={
                                          Object {
                                            "done": true,
                                            "importFunction": [Function],
                                            "load": [Function],
                                            "mark": Array [],
                                            "ok": true,
                                            "payload": Object {
                                              "process": [Function],
                                            },
                                            "promise": Promise {},
                                            "reset": [Function],
                                          }
                                        }
                                        render={[Function]}
                                        streamId={0}
                                      >
                                        <img
                                          className="backArrow"
                                          src={
                                            Object {
                                              "process": [Function],
                                            }
                                          }
                                        />
                                      </UnconnectedReactImportedComponent>
                                    </ReactImportedComponent>
                                  </Imported>
                                </Icon>
                                Back
                              </a>
                            </Link>
                          </Link>
                        </TextButton>
                      </ul>
                    </MenuList>
                  </main>
                </div>
              </Flex>
            </Box>
          </Card>
        </Menu>
        <div
          className="cards"
        >
          &lt;p&gt;This is a test&lt;/p&gt;
        </div>
      </div>
    </Flex>
  </Row>
</CardMenu>
`;

exports[`CardMenu should render correctly 2`] = `
<Card
  className="menu testClassString"
  id="123"
  title="Test"
>
  <MenuList
    button={
      Object {
        "backArrow": "arrow",
        "style": "blue",
        "text": "Back",
        "to": "/",
      }
    }
  >
    <Item
      active={true}
      exact={true}
      onClick={[Function]}
    >
      API Key
    </Item>
    <Item
      active={false}
      exact={true}
      onClick={[Function]}
    >
      Configure
    </Item>
    <div
      className="border"
    />
    <Item
      active={false}
      exact={true}
      onClick={[Function]}
    >
      Delete
    </Item>
  </MenuList>
</Card>
`;
